File: python/src/inspector/masking.py
"""Account-number masking, kept identical to ``go/internal/masking``.

The rule (ADR-011): mask account *numbers* to their last four characters, never
mask account *names*. Short identifiers are masked entirely, because revealing
"the last four" of a four-character value reveals all of it.

Two implementations of one rule is a risk, so ``tests/test_masking_parity.py``
asserts the Python output matches the Go output character for character on a
shared case table.
"""

from __future__ import annotations

import os

MIN_REVEALED = 4

_enabled = os.environ.get("HM_MASK_ACCOUNTS", "").lower() != "false"


def set_enabled(value: bool) -> None:
    """Override the environment. Intended for tests."""
    global _enabled
    _enabled = value


def enabled() -> bool:
    return _enabled


def account_number(value: str) -> str:
    """Mask all but the trailing four characters."""
    if not _enabled or not value:
        return value
    if len(value) <= MIN_REVEALED:
        return "*" * len(value)
    return "*" * (len(value) - MIN_REVEALED) + value[-MIN_REVEALED:]


def account_numbers(values: list[str]) -> list[str]:
    return [account_number(v) for v in values]


def account_name(value: str) -> str:
    """Deliberately a no-op, so the decision not to mask names is visible."""
    return value

File: go/internal/masking/masking.go
// Package masking redacts account numbers on the way out of the process.
//
// The rule (ADR-011): account *numbers* are masked to their last four
// characters; account *names* are never masked. Operators need to recognise a
// client at a glance without account numbers ending up in logs, screenshots or
// a shared terminal.
//
// The important design property is that masking happens in one place, at
// serialisation, rather than at each call site. A rule applied at call sites is
// a rule that will eventually be forgotten in the one code path that mattered.
package masking

import (
	"os"
	"strings"
	"sync/atomic"
)

// enabled is read on every call so tests can flip it without reconstructing
// anything. HM_MASK_ACCOUNTS=false is a local-debugging escape hatch only.
var enabled atomic.Bool

func init() {
	enabled.Store(strings.ToLower(os.Getenv("HM_MASK_ACCOUNTS")) != "false")
}

// SetEnabled overrides the environment. Intended for tests.
func SetEnabled(v bool) { enabled.Store(v) }

// Enabled reports whether masking is active.
func Enabled() bool { return enabled.Load() }

const minRevealed = 4

// AccountNumber masks all but the trailing four alphanumeric characters,
// preserving any separators so the shape of the identifier stays recognisable.
//
//	"8842-00119" -> "****-***119" is wrong; we keep it simple and predictable:
//	"8842-00119" -> "******0119"
//
// Short identifiers (four characters or fewer) are masked entirely rather than
// being returned in the clear, because revealing all of a short account number
// reveals the account.
func AccountNumber(s string) string {
	if !enabled.Load() || s == "" {
		return s
	}
	runes := []rune(s)
	if len(runes) <= minRevealed {
		return strings.Repeat("*", len(runes))
	}
	tail := runes[len(runes)-minRevealed:]
	return strings.Repeat("*", len(runes)-minRevealed) + string(tail)
}

// AccountNumbers masks a slice, returning a new slice.
func AccountNumbers(in []string) []string {
	if len(in) == 0 {
		return in
	}
	out := make([]string, len(in))
	for i, s := range in {
		out[i] = AccountNumber(s)
	}
	return out
}

// AccountName is the counterpart to AccountNumber and deliberately does
// nothing. It exists so that call sites read symmetrically and so that the
// decision not to mask names is visible in the code rather than implied by its
// absence.
func AccountName(s string) string { return s }
